Transaction Protocol Manager implementation for two party transaction negotiations

philipr-za opened the pull request:
## Description
The Transaction Protocol Manager implements a protocol to construct a Mimwblewimble transaction between two parties, a Sender and a Receiver. In this transaction the Sender is paying the Receiver from their inputs and also paying to as many change outputs as they like. The Receiver will receive a single output from this transaction. The module is written to allow for a single entity to construct a transaction by moving through the state machine and providing all the required data OR for two remote parties to collaborate to build the transaction by having the Transaction Protocol Manager construct and accept messages with the public information required from the two parties.

## Motivation and Context
Closes https://github.com/tari-project/tari/issues/42

## How Has This Been Tested?
An unit test is provided in the implementation

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [X] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [X] New Tests
* [ ] Documentation

## Checklist:
* [X] I'm merging against the `development` branch
* [X] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [X] I have added tests to cover my changes.

mikethetike commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+    InvalidStateError,
+    // An error occurred while performing a signature
+    #[error(no_from)]
+    SignatureError(SchnorrSignatureError),
```
Perhaps `SigningError` to differentiate it from `InvalidSignatureError`?

philipr-za replied:
Sure, done

mikethetike commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+//! |            +-------+                                           |            |
+//! |            |       | SetSenderNonce                            | Finalized  |
+//! |            +<------+                                           | Transaction|
+//! |            |                                                   |            |
```
Seems like these are more builder methods that could create the initial state. Can they be run in any order?

philipr-za replied:
Correct, they can be run in any order and do not leave that state so in that sense they are builder style methods.

philipr-za replied:
So after thinking about it bit more I do agree, I think I am going to make the initial setup (currently the SenderSetup state) into a builder and when it can be built starts the state machine. The Receiver will start the state machine by ingesting the public message from the sender. This will reduce the number of events significantly.

mikethetike commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+    /// are handled here. The previous state is consumed and dependant on the outcome of processing the event a new
+    /// state is returned.
+    fn handle_event(&mut self, event: TransactionProtocolEvent) -> Result<(), TransactionProtocolError> {
+        self.state = match self.state.clone() {
```
Seems strange to clone it here and then overwrite it. Could we perhaps use a `ref` in the match arms?

philipr-za replied:
So I tried many combo's but the issue is that the state changes consume the old state and give you a new one so I couldn't figure out how to allow that but still assign the new state to main struct without having move errors. It does feel a little redundant to make the state changes consume the old state if I am cloning here...

CjS77 replied:
Agree on the code smell here. you can have `&mut self` and then try burn the state in the arms.

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+/// in this state to accept the public data from the Receiver to advance to the next state.
+#[derive(Clone, Debug)]
+struct ReceiverOutputSetup {
+    state_data: TransactionProtocolStateData,
```
Box?

philipr-za replied:
See comment above (Vectors are already on the heap)

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+}
+
+impl ReceiverOutputSetup {
+    fn new(previous_state_data: TransactionProtocolStateData) -> ReceiverOutputSetup {
```
```suggestion
    fn new(previous_state_data: Box<TransactionProtocolStateData>) -> ReceiverOutputSetup {
```

philipr-za replied:
See comment above (Vectors are already on the heap)

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+    }
+
+    /// This method accepts data from a Sender and transitions into the next state
+    /// TODO Check that this is the Receiver and has no Sender state set?
```
Have you thought about having 2 different STNs? One for `Initiator` / Sender and one for `Receiver`? Then you don't have to do any checks here (and in the other places where you're having different transitions depending on the type of actor).

I'm also thinking forward to generalising this with 1 initiator and _n_ receivers; it might be easier to implement if the actors are separated.

philipr-za replied:
I did consider that but I also wanted to support this process being performed by a single actor (e.g. a wallet building a coin join) and it felt like a very subtle extension to allow for both use cases in a single state machine. However, I do agree with you that for future extension to ,multiple parties encoding this as two state-machines is a more modular design.

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+        }
+
+        self.state_data.receiver_private_nonce = Some(nonce.clone());
+        self.state_data.receiver_public_nonce = Some(PublicKey::from_secret_key(&nonce));
```
Swap the 2 lines above and you can drop the `clone`

philipr-za replied:
Yup, done

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+        self.state_data.receiver_partial_signature = Some(Signature::sign(
+            self.state_data.receiver_output_blinding_factor.unwrap(),
+            self.state_data.receiver_private_nonce.unwrap(),
+            challenge.clone(),
```
You don't need to `clone` here

philipr-za replied:
done

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+            return Err(TransactionProtocolError::IncompleteStateError);
+        }
+
+        let challenge = Challenge::<Blake256>::new()
```
Could this be DRYed up? It looks the same as the code in the previous function

philipr-za replied:
Made a function for it

CjS77 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+        // Validate final signature
+        if !self.state_data.final_signature.unwrap().verify_challenge(
+            &(&self.state_data.receiver_output_public_key.unwrap() + &self.state_data.sender_excess.unwrap()),
+            challenge.clone(),
```
You can drop this `clone`

philipr-za replied:
Done

philipr-za commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+    TransactionBuildError(TransactionError),
+}
+
+impl From<SchnorrSignatureError> for TransactionProtocolError {
```
I am glad you asked...to be honest now that I test it it seems to work fine not doing it explicitly but I remember having some issues with the embedded error type but ja seems to work perfectly without it.

Taken these out.

neonknight64 commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+    Failed(TransactionProtocolError),
+}
+
+/// This struct contains all the working data is required during the protocol. All fields are
```
```suggestion
/// This struct contains all the working data required during the protocol. All fields are
```

philipr-za replied:
Thanks, change made.

philipr-za commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+enum TransactionProtocolState {
+    SenderSetup(SenderSetup),
+    ReceiverOutputSetup(ReceiverOutputSetup),
+    ReceiverPartialSignature(ReceiverPartialSignature),
```
Changed the state machine names to be more descriptive.

philipr-za commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+    }
+
+    fn set_nonce(mut self, nonce: SecretKey) -> TransactionProtocolState {
+        self.state_data.sender_private_nonce = Some(nonce.clone());
```
Right you are, fixed

philipr-za commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+            sum = &sum + &o.commitment;
+        }
+        for i in self.state_data.inputs.clone() {
+            sum = &sum - &i.commitment;
```
done

philipr-za commented on base_layer/core/src/transaction_protocol_manager.rs:
```diff
@@ -0,0 +1,1046 @@
+/// transition to the next state.
+#[derive(Clone, Debug)]
+struct SenderSetup {
+    state_data: TransactionProtocolStateData,
```
I don't think this is needed as the only part of the structure that can get big are the vectors of inputs and outputs and I am pretty sure that Vector elements are always stored on the heap?

CjS77 replied:
You're right: https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees

Ok, it'll be fine then

philipr-za commented:
Thanks for the reviews, based on the comments I ended up doing a fairly large refactor. The two main things that I changed were:
- Split the single state machine into two separate ones, one for Sender and one for Receiver
- The initial Sender state is now built using a builder outside the state machines.
I think it is ready for another in depth review, I would ask that someone also do a proper review of the actual logic of the transaction protocol as I found a problem during this refactor which I missed so it definitely needs some scrutiny.

CjS77 commented:
I've brought this code into a separate set of PRs that implement TXs in a multiple recipient-ready way.
